refactor: Enhance ERC721VotingAdapterV1 with new token ID handling functions

adamgall opened the pull request:
This commit introduces several improvements to the `ERC721VotingAdapterV1` contract, including:

1. **New Functions:**
   - Added `_decodeTokenIds`, `_getUniqueTokenIds`, `_getOwnedTokenIds`, and `_getUnusedTokenIds` to streamline token ID processing and improve clarity.
   - Introduced `weightOfWithValidTokenIds` to return both weight and unused token IDs, enhancing the voting logic.

2. **Error Handling:**
   - Implemented new error messages for better feedback during voting operations, including checks for unowned and already used token IDs.

3. **Test Coverage:**
   - Expanded test cases to validate the new functionality, ensuring correct weight calculations and proper handling of edge cases.

These changes improve the maintainability and functionality of the voting adapter, ensuring a more robust voting process.

adamgall committed:
refactor: Enhance ERC721VotingAdapterV1 with new token ID handling functions

This commit introduces several improvements to the `ERC721VotingAdapterV1` contract, including:

1. **New Functions:**
   - Added `_decodeTokenIds`, `_getUniqueTokenIds`, `_getOwnedTokenIds`, and `_getUnusedTokenIds` to streamline token ID processing and improve clarity.
   - Introduced `weightOfWithValidTokenIds` to return both weight and unused token IDs, enhancing the voting logic.

2. **Error Handling:**
   - Implemented new error messages for better feedback during voting operations, including checks for unowned and already used token IDs.

3. **Test Coverage:**
   - Expanded test cases to validate the new functionality, ensuring correct weight calculations and proper handling of edge cases.

These changes improve the maintainability and functionality of the voting adapter, ensuring a more robust voting process.

tbwebb22 commented on contracts/deployables/strategies/adapters/ERC721VotingAdapterV1.sol:
```diff
@@ -51,90 +51,175 @@ contract ERC721VotingAdapterV1 is
-            return (new uint256[](0), 0);
+    function _getUniqueTokenIds(
+        uint256[] memory tokenIds
+    ) internal view virtual returns (uint256[] memory uniqueTokenIds) {
+        uint256[] memory tempUniqueTokenIds = new uint256[](tokenIds.length);
+        uint256 uniqueCount = 0;
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            bool isDuplicate = false;
+            for (uint256 j = 0; j < uniqueCount; j++) {
+                if (tempUniqueTokenIds[j] == tokenIds[i]) {
+                    isDuplicate = true;
+                    break;
+                }
+            }
+            if (!isDuplicate) {
+                tempUniqueTokenIds[uniqueCount] = tokenIds[i];
+                uniqueCount++;
+            }
         }
 
-        uint256[] memory tempValidTokenIds = new uint256[](tokenIds.length);
-        uint256 validCount = 0;
+        if (uniqueCount == 0) {
+            return new uint256[](0);
+        }
 
+        uniqueTokenIds = new uint256[](uniqueCount);
+        for (uint256 i = 0; i < uniqueCount; i++) {
+            uniqueTokenIds[i] = tempUniqueTokenIds[i];
+        }
+    }
```
We could simplify this to the bellow snippet - could also be applied to the other two similar internal view functions

```

function _getUniqueTokenIds(
        uint256[] memory tokenIds
    ) internal view virtual returns (uint256[] memory uniqueTokenIds) {
        // Create array with maximum possible size
        uniqueTokenIds = new uint256[](tokenIds.length);
        uint256 uniqueCount = 0;

        // Single loop to collect unique tokens
        for (uint256 i = 0; i < tokenIds.length; i++) {
            bool isDuplicate = false;
            for (uint256 j = 0; j < uniqueCount; j++) {
                if (uniqueTokenIds[j] == tokenIds[i]) {
                    isDuplicate = true;
                    break;
                }
            }
            if (!isDuplicate) {
                uniqueTokenIds[uniqueCount] = tokenIds[i];
                uniqueCount++;
            }
        }

        // Resize the array using assembly
        assembly {
            mstore(uniqueTokenIds, uniqueCount)
        }
    }
```

adamgall replied:
Thanks for this. Check the latest commit please

adamgall committed:
refactor: Optimize token ID handling in ERC721VotingAdapterV1

This commit refines the token ID processing logic in the `ERC721VotingAdapterV1` contract by eliminating temporary arrays and utilizing assembly for more efficient memory management. Key changes include:

1. **Function Improvements:**
   - Streamlined `_getUniqueTokenIds`, `_getOwnedTokenIds`, and `_getUnusedTokenIds` to directly populate the output arrays, enhancing performance and reducing memory overhead.

2. **Test Adjustments:**
   - Updated tests to reflect changes in token ownership logic, ensuring accurate weight calculations and correct filtering of owned and unused tokens.

These modifications enhance the efficiency and clarity of the voting adapter's token ID handling, contributing to a more robust voting process.
